src/serv_account/mod.rs

```diff
@@ -56,60 +49,125 @@ impl ServiceAccount {
-                    Ok(token) => token,
-                    Err(err) => return Err(err),
-                };
-
-                let expires_at = Utc::now().timestamp() as u64 + token.expires_in - 30;
-
-                self.access_token = Some(token.bearer_token());
-                self.expires_at = Some(expires_at);
-
-                Ok(token.bearer_token())
+                let new_token = self.get_fresh_access_token().await?;
+                *self.access_token.write().await = Some(new_token.clone());
+                Ok(new_token)
             }
         }
     }
 
-    async fn exchange_jwt_token_for_access_token(
-        &mut self,
-        jwt_token: jwt::JwtToken,
-    ) -> Result<Token> {
-        let req_builder = self.http_client.post(jwt_token.token_uri()).form(&[
-            ("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"),
-            ("assertion", &jwt_token.to_string()?),
-        ]);
-
-        let res = match req_builder.send().await {
-            Ok(resp) => resp,
-            Err(err) => return Err(ServiceAccountError::HttpReqwest(err)),
+    async fn get_fresh_access_token(&self) -> Result<AccessToken> {
+        let jwt_token = {
+            let mut token = jwt::JwtToken::from_key(&self.key)?.scope(self.scopes.clone());
+            if let Some(user_email) = &self.user_email {
+                token = token.sub(user_email.clone());
+            };
+            token
         };
 
-        let token = match res.json::<Token>().await {
-            Ok(token) => token,
-            Err(err) => return Err(ServiceAccountError::HttpReqwest(err)),
-        };
+        #[derive(Debug, Deserialize)]
+        pub struct TokenResponse {
+            token_type: String,
+            access_token: String,
+            expires_in: i64,
+        }
+
+        let response = self
+            .http_client
+            .post(jwt_token.token_uri())
+            .form(&[
+                ("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"),
+                ("assertion", &jwt_token.to_string()?),
+            ])
+            .send()
+            .await
+            .map_err(ServiceAccountError::HttpRequest)?;
+
+        if !response.status().is_success() {
+            return Err(ServiceAccountError::HttpRequestUnsuccessful(
+                response.status(),
+                response.text().await,
+            ));
+        }
+
+        let json = response
+            .json::<TokenResponse>()
+            .await
+            .map_err(ServiceAccountError::HttpJson)?;
+
+        if json.token_type != "Bearer" {
+            return Err(ServiceAccountError::AccessTokenNotBeaarer(json.token_type));
+        }
+
+        // Account for clock skew or time to receive or process the response
+        const LEEWAY: Duration = Duration::seconds(30);
 
-        Ok(token)
+        let expires_at = Utc::now() + Duration::seconds(json.expires_in) - LEEWAY;
```

Comment:
One of the options to resolve the issue with Duration above would be to keep it simple and use ints: 
`Utc::now().timestamp() as u64 + json.expires_in - 30`